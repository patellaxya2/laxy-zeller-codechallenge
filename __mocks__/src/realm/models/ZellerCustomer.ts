export const ZellerCustomer = jest.fn().mockImplementation(() => {
    return {
      id: "mock-id",
      fullName: "Mock User",
      email: "mock@email.com",
      role: "ADMIN",
    };
  });
  