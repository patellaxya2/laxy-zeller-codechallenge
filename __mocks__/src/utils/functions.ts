
const validateName = jest.fn();
const validateEmail = jest.fn();
export { validateEmail, validateName };
