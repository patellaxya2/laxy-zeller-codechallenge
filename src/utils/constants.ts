

export const roles = ["All","Admin", "Manager"]



