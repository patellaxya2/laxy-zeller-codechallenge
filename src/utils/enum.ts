export enum EROLES{
    ALL="ALL",
    ADMIN = "ADMIN",
    MANAGER="MANAGER"
    
}
