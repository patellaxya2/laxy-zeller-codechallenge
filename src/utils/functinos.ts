import Realm, { BSON } from "realm";
import { ZellerCustomer } from "../realm/models/ZellerCustomer";
import { IZellerCustomer } from "../types";

/**
 * Syncs backend data into Realm
 * @param realm The Realm instance
 * @param apiData The array of ZellerCustomer objects from backend
 */
export const syncZellerCustomers = async (
  realm: Realm,
  apiData: IZellerCustomer[]
) => {
  if (!apiData || apiData.length === 0) return;

  realm.write(() => {
    apiData.forEach((item) => {
      realm.create(
        ZellerCustomer,
        {
          id: item.id ,
          name: item.name,
          email: item.email,
          role: item.role, 
        },
        Realm.UpdateMode.Modified 
      );
    });
  });
};

export const validateName = (name: string) => {
  if (!name.trim()) return 'Name cannot be empty';  
  if (!/^[A-Za-z ]+$/.test(name)) return 'Only alphabets and spaces allowed';
  if (name.length > 50) return 'Name cannot exceed 50 characters';
  return '';
};

export const validateEmail = (email: string) => {
  if (!email) return '';
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (!emailRegex.test(email)) return 'Invalid email format';
  return '';
};


export const groupByLetter = (data: IZellerCustomer[]) => {
  const sorted = [...data].sort((a, b) => a.name.localeCompare(b.name));
  const grouped: { title: string; data: IZellerCustomer[] }[] = [];

  sorted.forEach((item) => {
      const letter = item.name[0].toUpperCase();
      const section = grouped.find((g) => g.title === letter);
      if (section) {
          section.data.push(item);
      } else {
          grouped.push({ title: letter, data: [item] });
      }
  });

  return grouped;
};