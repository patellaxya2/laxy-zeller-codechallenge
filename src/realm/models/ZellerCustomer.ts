// src/models/Task.ts

import { Realm, BSON, ObjectSchema } from "realm";
import { IZellerCustomer } from "../../types";
import { EROLES } from "../../utils/enum";

export class ZellerCustomer extends Realm.Object<IZellerCustomer> {
  id!: string;
  name!: string;
  email!: string;
  role!: EROLES;
    constructor(realm:Realm,name:string,email:string,role:EROLES) { 
        super(realm, { name, email, role })
    }
  
  static schema: ObjectSchema = {
    name: "ZellerCustomer",
    primaryKey: "id",
    properties: {
      id: {
        type: "string",
        default: () =>  Date.now()+""
      },
      name: "string",
      email: "string",
      role: "string",      
    },
  };
}