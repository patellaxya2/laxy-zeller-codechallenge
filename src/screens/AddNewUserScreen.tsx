import { NativeStackScreenProps } from '@react-navigation/native-stack';
import React, { useState } from 'react';
import { View, Text, TextInput, TouchableOpacity, StyleSheet, Animated, Alert } from 'react-native';
import { StackParamList } from '../types';
import { EROLES } from '../utils/enum';
import { ZellerCustomer } from '../realm/models/ZellerCustomer';
import { useRealm } from '../realm/setup';
import theme from '../styles/theme';
import useTabAnimation from '../hooks/useTabAnimation';
import AnimatedTabView from '../components/AnimatedTabView';
import { validateEmail, validateName } from '../utils/functinos';
import { SafeAreaView } from 'react-native-safe-area-context';
import Loader from '../components/Loader';

type AddNewUserScreenProps = NativeStackScreenProps<StackParamList, 'AddNewUser'>;
const roles = ["Admin", "Manager"]
const AddNewUserScreen: React.FC<AddNewUserScreenProps> = ({ navigation }) => {
    const realm = useRealm();
    const { translateX, startAnimation, tabWidth } = useTabAnimation({ items: roles });

    const [fullName, setFullName] = useState('');
    const [email, setEmail] = useState('');
    const [userRole, setUserRole] = useState<string>(roles[0]);
    const [loader, setLoader] = useState(false)

    const handleCreate = () => {
        const emailError = validateEmail(email);
        const fullNameError = validateName(fullName);
        if (fullNameError) {
            Alert.alert("Error", fullNameError)
        }
        else if (emailError) {
            Alert.alert("Error", emailError)
        }
        else {
            setLoader(true)
            try {

                realm.write(() => {
                    return new ZellerCustomer(realm, fullName, email, userRole === "Admin" ? EROLES.ADMIN : EROLES.MANAGER);
                })
                // console.log("successfully saved!");
                setLoader(false)

                Alert.alert("Success", "Data saved successfully.")
            } catch (error) {
                setLoader(false)

                // console.log("Something went wrong!", error);
                Alert.alert("Error", "Something went wrong!")

            }
        }
    };
    const handleTabPress = (role: string, index: number) => {
        setUserRole(role);
        startAnimation(index, 300);
    };
    return (

        <SafeAreaView style={styles.modalContainer}>
            <Loader visible={loader} />
            <View style={styles.modalContent}>

                <TouchableOpacity style={styles.closeButton} onPress={() => {
                    navigation.goBack()
                }}>
                    <Text style={styles.closeButtonText}>x</Text>
                </TouchableOpacity>
                <Text style={styles.title}>New User</Text>

                <TextInput
                    style={styles.input}
                    placeholder="Full Name*"
                    value={fullName}
                    onChangeText={setFullName}
                />
                <TextInput
                    style={styles.input}
                    placeholder="Email"
                    value={email}
                    onChangeText={setEmail}
                    keyboardType="email-address"
                    autoCapitalize="none"
                />

                <Text style={styles.label}>User Role</Text>


            </View>


            {/* <View style={styles.roleContainer}> */}
            <AnimatedTabView
                items={roles}
                handleTabPress={handleTabPress}
                selected={userRole}
                tabWidth={tabWidth}
                translateX={translateX as Animated.Value}
            />
            <View style={{ padding: 20, flex: 1, justifyContent: "flex-end" }}>
                <TouchableOpacity style={styles.createButton} onPress={handleCreate}>
                    <Text style={styles.createButtonText}>Create User</Text>
                </TouchableOpacity>
            </View>
            {/* </View> */}
        </SafeAreaView>


    )
}

export default AddNewUserScreen

const styles = StyleSheet.create({
    modalContainer: {
        flex: 1,
        backgroundColor: theme.colors.primary,

    },
    modalContent: {
        backgroundColor: theme.colors.primary,
        paddingHorizontal: 20,
        borderRadius: 10,
        width: '100%',
        // height: "100%",

    },
    title: {
        fontSize: 20,
        fontWeight: 'bold',
        marginVertical: 20,
    },
    input: {
        width: '100%',
        borderBottomWidth: 1,
        borderColor: theme.colors.light_gray,
        padding: 10,
        marginVertical: 15,
    },
    label: {
        marginVertical: 20,

    },
    roleContainer: {
        flex: 1,
        width: "100%"
    },
    roleButton: {
        padding: 10,
        borderWidth: 1,
        borderColor: theme.colors.light_gray,
        borderRadius: 15,
        flex: 1,
        marginHorizontal: 5,
    },
    selectedRole: {
        backgroundColor: theme.colors.secondary,
    },
    roleText: {
        textAlign: 'center',
    },
    createButton: {
        backgroundColor: theme.colors.secondary,
        padding: 15,
        borderRadius: 20,
        width: '100%',
        alignItems: 'center',
    },
    createButtonText: {
        color: theme.colors.primary,
        fontWeight: 'bold',
    },
    closeButton: {
        paddingLeft: 10
    },
    closeButtonText: {
        fontSize: 35,
        color: theme.colors.secondary,
    },
});
