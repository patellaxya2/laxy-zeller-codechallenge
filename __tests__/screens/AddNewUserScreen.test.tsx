

import { fireEvent, render, waitFor } from "@testing-library/react-native";
import AddNewUserScreen from "../../src/screens/AddNewUserScreen";
import { Alert } from "react-native";
import { validateEmail, validateName } from "../../src/utils/functinos";
import { EROLES } from "../../src/utils/enum";
import { ZellerCustomer } from "../../src/realm/models/ZellerCustomer";
import { act } from "react";

const mockStartAnimation = jest.fn()
jest.mock("../../src/hooks/useTabAnimation", () => {
    return jest.fn(() => ({
        startAnimation: mockStartAnimation
    }));
});
jest.mock('../../src/utils/functinos')

jest.spyOn(Alert, "alert").mockImplementation(() => { });
jest.mock("../../src/realm/models/ZellerCustomer");

const mockGoBack = jest.fn();
const navigation = { goBack: mockGoBack };
const mockValidateName = validateName as jest.Mock;
const mockValidateEmail = validateEmail as jest.Mock;
const mockWrite = jest.fn((cb) => cb());


jest.mock("../../src/realm/setup", () => ({
    useRealm: () => ({
        write: mockWrite
    })
}))

describe('AddNewUserScreen', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it("renders correctly", () => {
        const { getByText, getByPlaceholderText } = render(<AddNewUserScreen navigation={navigation as any} route={{} as any} />);
        expect(getByText("New User")).toBeTruthy();
        expect(getByPlaceholderText("Full Name*")).toBeTruthy();
        expect(getByPlaceholderText("Email")).toBeTruthy();
        expect(getByText("Create User")).toBeTruthy();
    })


    it("shows error when email is invalid", () => {
        const { getByText, getByPlaceholderText } = render(
            <AddNewUserScreen navigation={navigation as any} route={{} as any} />
        );

        mockValidateName.mockReturnValue(null);
        mockValidateEmail.mockReturnValue("Invalid email");

        fireEvent.changeText(getByPlaceholderText("Full Name*"), "John Doe");
        fireEvent.changeText(getByPlaceholderText("Email"), "wrong");
        fireEvent.press(getByText("Create User"));

        expect(Alert.alert).toHaveBeenCalledWith("Error", "Invalid email");
    });


    it("shows error when fullName is invalid", () => {
        const { getByText, getByPlaceholderText } = render(
            <AddNewUserScreen navigation={navigation as any} route={{} as any} />
        );
        const mockValidateName = validateName as jest.Mock;
        const mockValidateEmail = validateEmail as jest.Mock;
        mockValidateName.mockReturnValue("Invalid name");
        mockValidateEmail.mockReturnValue(null);

        fireEvent.changeText(getByPlaceholderText("Full Name*"), "John32@Doe");
        fireEvent.changeText(getByPlaceholderText("Email"), "john@gmail.com");
        fireEvent.press(getByText("Create User"));

        expect(Alert.alert).toHaveBeenCalledWith("Error", "Invalid name");
    });

    it("saves user successfully when inputs are valid", async () => {
        const { getByText, getByPlaceholderText } = render(
            <AddNewUserScreen navigation={navigation as any} route={{} as any} />
        );
        mockValidateName.mockReturnValue("");
        mockValidateEmail.mockReturnValue("");
        fireEvent.changeText(getByPlaceholderText("Full Name*"), "Laxy Doe");
        fireEvent.changeText(getByPlaceholderText("Email"), "Laxy@test.com");
        fireEvent.press(getByText("Create User"));

        await waitFor(() => {
            expect(mockWrite).toHaveBeenCalled();
            expect(ZellerCustomer).toHaveBeenCalledWith(
                expect.anything(), // ignore realm object
                "Laxy Doe",
                "Laxy@test.com",
                EROLES.ADMIN,

                // default role is Admin
            );
            expect(Alert.alert).toHaveBeenCalledWith(
                "Success",
                "Data saved successfully."
            );
        })

    })


    it("handles realm write error", async () => {
        mockWrite.mockImplementationOnce(() => {
            throw new Error("Realm error");
        });

        mockValidateName.mockReturnValue("");
        mockValidateEmail.mockReturnValue("");

        const { getByText, getByPlaceholderText } = render(
            <AddNewUserScreen navigation={navigation as any} route={{} as any} />
        );

        fireEvent.changeText(getByPlaceholderText("Full Name*"), "Jane Doe");
        fireEvent.changeText(getByPlaceholderText("Email"), "jane@test.com");
        fireEvent.press(getByText("Create User"));

        await waitFor(() => {
            expect(Alert.alert).toHaveBeenCalledWith("Error", "Something went wrong!");
        });
    });

    it("switches role via AnimatedTabView", async () => {
        const { getByTestId } = render(
            <AddNewUserScreen navigation={navigation as any} route={{} as any} />
        );

        act(() => {
            fireEvent.press(getByTestId("tab-Manager"));
        });

        await waitFor(() => {

            expect(mockStartAnimation).toHaveBeenCalledWith(1, 300);
        })
    });

    test('Navigation go back should call', () => {
        const { getByText } = render(
            <AddNewUserScreen navigation={navigation as any} route={{} as any} />
        );
        fireEvent.press(getByText("x"));
        expect(navigation.goBack).toHaveBeenCalled();

    })


})