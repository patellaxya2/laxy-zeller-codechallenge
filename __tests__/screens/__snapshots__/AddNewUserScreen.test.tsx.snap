// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot should Snapshot match 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#FFF",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#FFF",
        "borderRadius": 10,
        "paddingHorizontal": 20,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "paddingLeft": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "#007aff",
            "fontSize": 35,
          }
        }
      >
        x
      </Text>
    </View>
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "marginVertical": 20,
        }
      }
    >
      New User
    </Text>
    <TextInput
      onChangeText={[Function]}
      placeholder="Full Name*"
      style={
        {
          "borderBottomWidth": 1,
          "borderColor": "#ccc",
          "marginVertical": 15,
          "padding": 10,
          "width": "100%",
        }
      }
      value=""
    />
    <TextInput
      autoCapitalize="none"
      keyboardType="email-address"
      onChangeText={[Function]}
      placeholder="Email"
      style={
        {
          "borderBottomWidth": 1,
          "borderColor": "#ccc",
          "marginVertical": 15,
          "padding": 10,
          "width": "100%",
        }
      }
      value=""
    />
    <Text
      style={
        {
          "marginVertical": 20,
        }
      }
    >
      User Role
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "margin": 10,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#F0F2F5",
          "borderRadius": 20,
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "center",
          "position": "relative",
          "width": "80%",
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "borderColor": "#007aff",
            "borderRadius": 20,
            "borderWidth": 1,
            "height": "100%",
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateX": undefined,
              },
            ],
            "width": undefined,
            "zIndex": 1,
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "opacity": 1,
          }
        }
        testID="tab-Admin"
      >
        <Text
          style={
            [
              {
                "color": "#333",
                "fontSize": 14,
              },
              {
                "color": "#007aff",
                "fontWeight": "600",
              },
            ]
          }
        >
          Admin
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "opacity": 1,
          }
        }
        testID="tab-Manager"
      >
        <Text
          style={
            [
              {
                "color": "#333",
                "fontSize": 14,
              },
              false,
            ]
          }
        >
          Manager
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flex": 1,
        "justifyContent": "flex-end",
        "padding": 20,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#007aff",
          "borderRadius": 20,
          "opacity": 1,
          "padding": 15,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontWeight": "bold",
          }
        }
      >
        Create User
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
